src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs

```diff
@@ -185,53 +199,142 @@ private void GenerateIAsyncEnumeratorImplementation_MoveNextAsync()
                     F.WellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerator_T)
                     .Construct(_currentField.Type.TypeSymbol);
 
-                MethodSymbol IAsyncEnumerableOfElementType_MoveNextAsync =
-                    F.WellKnownMethod(WellKnownMember.System_Collections_Generic_IAsyncEnumerator_T__MoveNextAsync)
+                MethodSymbol IAsyncEnumerableOfElementType_MoveNextAsync = F.WellKnownMethod(WellKnownMember.System_Collections_Generic_IAsyncEnumerator_T__MoveNextAsync)
                     .AsMember(IAsyncEnumeratorOfElementType);
 
                 // The implementation doesn't depend on the method body of the iterator method.
                 OpenMethodImplementation(IAsyncEnumerableOfElementType_MoveNextAsync, hasMethodBodyDependency: false);
 
-                var ifFinished = F.If(
-                    // if (State == StateMachineStates.FinishedStateMachine)
+                TypeSymbol returnType = IAsyncEnumerableOfElementType_MoveNextAsync.ReturnType.TypeSymbol;
+
+                GetPartsForStartingMachine(returnType,
```

Comment:
>GetPartsForStartingMachine [](start = 16, length = 26)

:memo: extracted a number of locals to this helper method, so that it could be re-used in implementation of `DisposeAsync()`.